keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+**Note:** namespaced OBs is a change from the original design and was suggested at the 2019 NA Kubecon storage face-to-face with the premise being that if there is no _technical_ reason for a cluster scoped resource then it should be namespaced.
+
+As is true for PVCs-PVs, there is a 1:1 relationship between an OBC and an OB, and as will be seen below, there is a [_binding_](#binding) between an OBC and OB. 
+
+OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, bucket lifecycle policies, etc.
+
+### Moving Parts
+
+Before deploying an object bucket consuming app, the OB and OBC CRDs need to be created with RBAC rules granting _get_, _list_, _watch_, and _delete_ verbs.
+**Note:** OBCs do not need delete permission but OBs, secrets, and configmaps do.
+An object store needs to be created and referenced in the storage classes used by OBCs.
+The OBCs needed by the app must exist in the app's namespace.
+
+As is true for dynamic PV provisioning, a bucket provisioner pod needs to be running for each type of object-store existing in the Kubernetes cluster.
+For example, for AWS S3, the developer creates an OBC referencing a storage class which references the S3 store.
+This storage class needs to be created by the admin.
+The S3 provisioner pod watches for OBCs whose storage classes point to the AWS S3 provisioner, while ignoring all other OBCs.
+Likewise, the same cluster can also run the Rook-Ceph RGW provisioner, which also watches OBCs, only handling OBCs that reference storage classes which define ceph-rgw.
+
+**Note:** it is possible for one provisioner to handle OBCs for different instances of the same type of object store.
+
+#### Greenfield (new)
+
+An OBC for a new bucket is defined by its inclusion of a generated or static bucket name.
+In this case the referenced storage class omits a bucket name.
+A new bucket OBC triggers the correct provisioner to create a new bucket and the associated artifacts (credentials, policy, etc). 
+The library, in turn, creates a secret, configmap, and OB based on information returned by the provisioner.
+
+The secret contains the credentails needed to access the bucket and the configmap contains bucket endpoint information.
+**Note:** it was mentioned at the 2019 NA Kubecon storage face-to-face that to improve scaling we could decide to move the configmap info into the secret, thus reducing k8s resources required.
+The OB is an abstraction of the bucket and can save some state data needed by provisioners.
+The secret and configmap reside in the same namespace as the OBC, and the OB lives in the provisioner's namespace.
+The app pod consumes the secret and configmap and thus can reference the bucket.
```

Comment:
Multiple client pods can consume the same OBC to read/write to the same bucket, right?

Reply:
yes, subject to object store access rules. k8s has no opinion on this since there is no accessMode defined for buckets.

Reply:
Is there any reason why we don't want accessMode input(opinion) from K8s side, presumably in claim? Read-only and Read-write access to buckets seem to be reasonable use cases. I'm imagining in a brownfield scenario, user can have a bucket setup for both consumer and producer application, accessMode would be handy to limit consumer to read-only access and allow producer to have read write access. Also read-only vs read-write seem to be well-supported across major object storage vendors.

Reply:
Agree that there are good use cases for `accessMode`. My (admittedly dated) understanding is that accessMode is unenforceable by k8s. Is this still true @jsafrane?
For bucket provisioning, the underlying object store would need to support various access modes.

Reply:
Yes, AccessMode is used only during binding / provisioning, but it's never used when running a pod - you can have multiple pods using the same RWO volume in parallel.